client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,933 @@
+	if c.token.Root && c.token.TTL == 0 {
+		c.logger.Printf("[DEBUG] client/vaultclient: not renewing token as it is a non-expiring root token")
+	} else {
+		c.logger.Printf("[DEBUG] client/vaultclient: token lease duration is %v", time.Duration(c.token.CreationTTL)*time.Second)
+
+		// Renew the token and place it in renewal min-heap
+		errCh := c.RenewToken(c.config.Token, c.token.CreationTTL)
+
+		// Catch the renewal error of VaultClient's token.
+		go func(errCh <-chan error) {
+			var err error
+			for {
+				select {
+				case err = <-errCh:
+					c.logger.Printf("[ERR] client/vaultclient: error while renewing the vault client's token: %v", err)
+				}
+			}
+		}(errCh)
+	}
+}
+
+// parseSelfToken looks up the VaultClient's token in vault and parses its data
+// storing it in the client. If the token is not valid for Nomads purposes an
+// error is returned.
+func (c *vaultClient) parseSelfToken() error {
+	// Get the initial lease duration
+	auth := c.client.Auth().Token()
+	self, err := auth.LookupSelf()
+	if err != nil {
+		return fmt.Errorf("failed to lookup VaultClient's token: %v", err)
+	}
+
+	// Read and parse the fields
+	var data tokenData
+	if err := mapstructure.WeakDecode(self.Data, &data); err != nil {
+		return fmt.Errorf("failed to parse Vault token's data block: %v", err)
+	}
+
+	root := false
+	for _, p := range data.Policies {
+		if p == "root" {
+			root = true
+			break
+		}
+	}
+
+	if !data.Renewable && !root {
+		return fmt.Errorf("vault token is not renewable or root")
+	}
+
+	if data.CreationTTL == 0 && !root {
+		return fmt.Errorf("invalid lease duration of zero")
+	}
+
+	if data.TTL == 0 && !root {
+		return fmt.Errorf("token TTL is zero")
+	}
+
+	if !root && data.Role == "" {
+		return fmt.Errorf("token role name must be set when not using a root token")
+	}
+
+	data.Root = root
+	c.token = &data
+	return nil
+}
+
+// Stops the renewal loop of vault client
+func (c *vaultClient) Stop() {
+	if !c.config.Enabled || !c.running {
+		return
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.running = false
+	close(c.stopCh)
+}
+
+// DeriveToken takes in an allocation and a set of tasks and for each of the
+// task, it derives a vault token from nomad server and unwraps it using vault.
+// The return value is a map containing all the unwrapped tokens indexed by the
+// task name.
+func (c *vaultClient) DeriveToken(alloc *structs.Allocation, taskNames []string) (map[string]string, error) {
+	var result *multierror.Error
+
+	if !c.running {
+		result = multierror.Append(fmt.Errorf("vault client is not running"))
+		return nil, result
+	}
+
+	if alloc == nil {
+		result = multierror.Append(fmt.Errorf("nil allocation"))
+		return nil, result
+	}
+	if taskNames == nil || len(taskNames) == 0 {
+		result = multierror.Append(fmt.Errorf("missing task names"))
+		return nil, result
+	}
+
+	found := false
+	verifiedTasks := []string{}
+	// Check if the given task names actually exist in the allocation
```

Comment:
The allocation has a field called `TaskGroup`. We can use that to index into the `alloc.Job.TaskGroups` and then go through the list